### PR TITLE
Key management

### PR DESCRIPTION
**Pull request by tajhlande**
[//]: # "Please read `CONTRIBUTING.md` before opening a pull request."

# Key Management updates
[//]: # "Describe the changes in this pull request here."

Added methods for key management functions, particularly to support namespaced keys:

- `cache.get_all_keys()` returns a list of all keys
- `cache.find_keys_starting_with(pattern)` returns a list of keys starting with the pattern
- `cache.find_matching_keys(pattern)` returns a list of keys containing the pattern anywhere
- `cache.clear_keys_starting_with(pattern)` clears all keys starting with the pattern
- `cache.clear_matching_keys(pattern)` clears all keys containing the pattern anywhere

### CODE REVIEW

**Comment by tajhlande** [comment]
It's not clear to me if I need to do any additional work on documentation or if that will be handled by `mkdocstrings` when you do a release.  If I need to do something else, please let me know.

File: sqlite3_cache/cache.py [hunk 1]
```diff
@@ -522,3 +525,99 @@ def ttl_many(self, keys: List[str]) -> Dict[str, int]:
             self._con.commit()
 
         return results
+
+    def _filter_key_result_list(self, unfiltered: List[Tuple[str, Any]]) -> List[str]:
+        """
+        Filters key result list to only those keys for cache items that are still alive,
+        and purges expired items from the cache
+        :param unfiltered: A list of key and expiration tuples
```
Comment (MrThearMan) [inline sqlite3_cache/cache.py:533]: Space before `:params` (should be a ruff lint rule, don't know why it didn't catch this).

File: sqlite3_cache/cache.py [hunk 2]
```diff
@@ -522,3 +525,99 @@ def ttl_many(self, keys: List[str]) -> Dict[str, int]:
...
+        """
+        Find keys that start with the given pattern.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+        Will only return keys that exist in the cache for currently valid cache items.
+
+        :param pattern: The pattern to match at the start of the key.
+        :return: List of matching cache keys in sort order.
+        """
+        data = {"pattern": f"{''}{pattern}%"}
```
Comment (MrThearMan) [inline sqlite3_cache/cache.py:575]: `{''}` seems unnecessary here.

File: sqlite3_cache/cache.py [hunk 3]
```diff
@@ -522,3 +525,99 @@ def ttl_many(self, keys: List[str]) -> Dict[str, int]:
...
+        """
+        Find keys that contain the given pattern anywhere in the string.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+        Will only return keys that exist in the cache for currently valid cache items.
+
+        :param pattern: The pattern to find in matching keys.
+        :return: List of matching cache keys in sort order.
+        """
+        data = {"pattern": f"{'%'}{pattern}%"}
```
Comment (MrThearMan) [inline sqlite3_cache/cache.py:593]: Pattern should probably be `f"%{pattern}%"`. Alternatiely, you could leave the `%`'s out and make this a more generic method, and refactor the "starting_with" method to use it.

File: sqlite3_cache/cache.py [hunk 4]
```diff
@@ -522,3 +525,99 @@ def ttl_many(self, keys: List[str]) -> Dict[str, int]:
...
+
+    def clear_keys_starting_with(self, pattern: str) -> None:
+        """
+        Clear all keys from the cache that start with the given pattern.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+
+        :param pattern: The pattern to match at the start of the key.
+        """
+        data = {"pattern": f"{''}{pattern}%"}
```
Comment (MrThearMan) [inline sqlite3_cache/cache.py:609]: {''} seems unnecessary here.

File: sqlite3_cache/cache.py [hunk 5]
```diff
@@ -522,3 +525,99 @@ def ttl_many(self, keys: List[str]) -> Dict[str, int]:
...
+
+    def clear_matching_keys(self, pattern: str) -> None:
+        """
+        Clear all keys from the cache that contain the given pattern anywhere in the string.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+
+        :param pattern: The pattern to find in matching keys.
+        """
+        data = {"pattern": f"{'%'}{pattern}%"}
```
Comment (MrThearMan) [inline sqlite3_cache/cache.py:621]: Pattern should probably be `f"%{pattern}%"`. Alternatiely, you could leave the `%`'s out and make this a more generic method, and refactor the "starting_with" method to use it.

**Review by MrThearMan — CHANGES_REQUESTED** [review]
Looks great! Matches my style well, great tests, and I only found a few minor changes 😊

Documentation needs to be duplicated to the `/docs/ref.md ` file. Still, thanks for including it as dosctrings as well!

**Comment by MrThearMan** [comment]
Hmm, not sure why tests pipelines are not triggering 🤔 I might try [closing and opening the PR](https://stackoverflow.com/a/67986789) and see if that fixes things if they don't start working soon.

File: sqlite3_cache/cache.py (hunk 2 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py:575]: I jumped through hoops to reuse a SQL string and avoid linting complaints about a boolean flag, and then this was left behind from some refactoring. I can just make two SQL strings and all of this will become simpler.

File: sqlite3_cache/cache.py (hunk 5 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py:621]: it was hard for me to find a way to do this that both felt like the simplest clearest API to declare and also passed linting.

File: sqlite3_cache/cache.py (hunk 4 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py]: Fixed

File: sqlite3_cache/cache.py (hunk 3 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py]: Fixed

File: sqlite3_cache/cache.py (hunk 2 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py]: Fixed

File: sqlite3_cache/cache.py (hunk 1 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py:533]: Fixed

File: sqlite3_cache/cache.py (hunk 5 above)
Comment (tajhlande) [reply sqlite3_cache/cache.py]: Fixed

**Comment by tajhlande** [comment]
I think I've addressed all your concerns.  I will leave the conversations to you for resolution.

File: sqlite3_cache/cache.py (hunk 1 above)
Comment (MrThearMan) [reply sqlite3_cache/cache.py:533]: Sorry, I wasn't precise enough: There _**should**_ be a newline between the body and the params. Could you add them back?

Comment (tajhlande) [reply sqlite3_cache/cache.py:534]: and now fixed the proper way to add space where it was missing

File: sqlite3_cache/cache.py (hunk 5 above)
Comment (MrThearMan) [reply sqlite3_cache/cache.py]: I think you could do something like this:

```python
def find_matching_keys(self, pattern: str) -> List[str]:
    # Any custom pattern can be used here
    data = {"pattern": pattern}
    fetched: List[Tuple[str, Any]] = self._con.execute(self._find_matching_keys_sql, data).fetchall()
    if not fetched:
        return []

    return self._filter_key_result_list(fetched)

def find_keys_keys_starting_with(self, pattern: str) -> List[str]:
    return self.find_matching_keys(f"{pattern}%")

def find_keys_keys_ending_with(self, pattern: str) -> List[str]:
    return self.find_matching_keys(f"%{pattern}")

def find_keys_keys_containing(self, pattern: str) -> List[str]:
    return self.find_matching_keys(f"%{pattern}%")
```

Comment (tajhlande) [reply sqlite3_cache/cache.py]: I didn't think you'd want to expose the SQL `%` syntax of the `LIKE` operator in the API itself.  But now that I know it's OK with you, it will simplify things!

Comment (tajhlande) [reply sqlite3_cache/cache.py]: OK, I took your suggestion and rewrote the methods this way, including for the `clear_keys` methods, and updated `docs/ref.md` as well.

**Comment by tajhlande** [comment]
Looks like the PR needs you to approve running the tests

**Comment by MrThearMan** [comment]
Seems like coveralls is unhappy that the cases where `get_all_keys`and `find_matching_keys` return an empty list are not covered. Could you add these to the test suite?

**Review by MrThearMan — APPROVED** [review]
LGTM, just add the test cases for empty lists from `get_all_keys` and `find_matching_keys` and we're golden 👍

**Comment by tajhlande** [comment]
I don't know how to run those additional workflows on my own, so hopefully this commit fixes things for coveralls.

**Review by MrThearMan — APPROVED** [review]

**Comment by MrThearMan** [comment]
Released in [0.2.1](https://github.com/MrThearMan/sqlite3-cache/releases/tag/v0.2.1).

**Comment by tajhlande** [comment]
Thank you!
